experiment/builder_runner.py

```diff
@@ -322,21 +322,60 @@ def get_coverage_local(
 
 class CloudBuilderRunner(BuilderRunner):
   """Cloud BuilderRunner."""
-  _RETRYABLE_ERRORS = [
-      # As mentioned in pr #100.
-      'RESOURCE_EXHAUSTED',
-      # Temp workaround for issue #12.
-      'You do not currently have an active account selected',
-      # Workaround for issue #85.
-      'gcloud crashed (OSError): unexpected end of data',
-  ]
 
   def __init__(self, *args, experiment_name: str, experiment_bucket: str,
                **kwargs):
     self.experiment_name = experiment_name
     self.experiment_bucket = experiment_bucket
     super().__init__(*args, **kwargs)
 
+  @staticmethod
+  def _run_with_retry_control(target_path: str, *args, **kwargs):
+    """sp.run() with controllable retry and customized exponential backoff."""
+    # List of (error_str, exp_backoff_func).
+    retryable_errors = [
+        # As mentioned in pr #100.
+        ('RESOURCE_EXHAUSTED', lambda x: 5 * 2**x + random.randint(50, 90)),
+        # As mentioned in pr $151.
+        ('BrokenPipeError: [Errno 32] Broken pipe',
+         lambda x: 5 * 2**x + random.randint(1, 5)),
+        # Temp workaround for issue #12.
+        ('You do not currently have an active account selected',
+         lambda x: 5 * 2**x),
+        # Workaround for issue #85.
+        ('gcloud crashed (OSError): unexpected end of data', lambda x: 5 * 2**x
+        ),
+    ]
+
+    for attempt_id in range(1, CLOUD_EXP_MAX_ATTEMPT + 1):
+      try:
+        sp.run(*args, capture_output=True, check=True, **kwargs)
+        return
+      except sp.CalledProcessError as e:
+        # Replace \n for single log entry on cloud.
+        stdout = e.stdout.decode('utf-8').replace('\n', '\t')
+        stderr = e.stderr.decode('utf-8').replace('\n', '\t')
+
+        delay = 0
+        for err, delay_f in retryable_errors:
+          if err in stdout + stderr:
+            delay = delay_f(attempt_id)
+            break
+
+        if attempt_id == CLOUD_EXP_MAX_ATTEMPT:
+          delay = 0
+
+        if delay:
+          logging.warning(
+              'Failed to evaluate %s on cloud, attempt %d, retry in %ds:\n'
+              '%s\n%s', os.path.realpath(target_path), attempt_id, delay,
+              stdout, stderr)
+          time.sleep(delay)
+          continue
+        logging.error('Failed to evaluate %s on cloud, attempt %d:\n%s\n%s',
+                      os.path.realpath(target_path), attempt_id, stdout, stderr)
+        raise e
```

Comment:
I guess you added `if attempt_id == CLOUD_EXP_MAX_ATTEMPT:` because you'd like to identify the final failure after `CLOUD_EXP_MAX_ATTEMPT` attempts.
Would it be clearer to move the final failure logging outside the `except`?

For example, here we have:
```python
for attempt_id in range():
  try:
    sp.run()
    return
  except sp.CalledProcessError as e:
  ... decode stdout/stderr...
  delay = 0
  for err, delay_f in retryable_errors:
    if err in stdout + stderr:
      delay = delay_f(attempt_id)
      break
    logging.warning(...)
    time.sleep(delay)
logging.error(... )
raise e
```

Reply:
We'd still want to log stdout and stderr from the exception.
The last error logging will handle two situations:
1. Exceeding max attempt;
2. Non retry-able error occurred.
Logging the attempt_id can also help distinguish between the two.

What we can do alternatively might be to remove `check=True` and handle process error checking manually without a try-except block

Reply:
> Non retry-able error occurred.

I see. Should we have a `break` after `logging.error()` to avoid going through another iteration?

And I presume `if attempt_id == CLOUD_EXP_MAX_ATTEMPT` is not needed either way, because you have initialised `delay = 0` before that?

Reply:
Also you can access `attempt_id` after the for loop terminates.
It will be the value used in the last iteration.

Reply:
`break` added and max attempt checking merged with error checking

> access `attempt_id` after the for loop terminates

Ah good to know, confusing [scope in python](https://mail.python.org/pipermail/python-ideas/2008-October/002109.html) :(

Reply:
Should look better now with [814580f](https://github.com/google/oss-fuzz-gen/pull/154/commits/814580fd49b09696352dfeb5e037c6a63cc9f303) I think